Content.Shared/Turnstile/Systems/SharedTurnstileSystem.cs

```diff
@@ -0,0 +1,230 @@
+
+                // We have to set collidable to false for one frame, otherwise the client does not
+                // predict the removal of the contacts properly, and instead thinks the mob is colliding when it isn't.
+                SetCollidable(ent, false);
+                SetState(ent, TurnstileState.Rotating);
+            }
+            else
+            {
+                // Reject the entity, play sound with cooldown
+                Audio.PlayPvs(turnstile.BumpSound, ent.Owner, AudioParams.Default.WithVolume(-3));
+            }
+        }
+        else
+        {
+            // Reject the entity, play sound with cooldown
+            Audio.PlayPvs(turnstile.BumpSound, ent.Owner, AudioParams.Default.WithVolume(-3));
+        }
+    }
+
+    /// <summary>
+    ///     Makes a turnstile proceed to the next state.
+    /// </summary>
+    private void NextState(Entity<TurnstileComponent> ent, TimeSpan time)
+    {
+        var turnstile = ent.Comp;
+
+        if (turnstile.State == TurnstileState.Rotating)
+        {
+            // Check and see if we're still colliding with the admitted entity.
+
+            var stillCollidingWithAdmitted = GetIsCollidingWithAdmitted(ent);
+            if (!stillCollidingWithAdmitted)
+            {
+                turnstile.NextStateChange = null;
+                SetState(ent, TurnstileState.Idle);
+                RemComp<PreventCollideComponent>(ent.Owner);
+            }
+            else
+            {
+                turnstile.NextStateChange = GameTiming.CurTime + turnstile.TurnstileTurnTime;
+            }
+        }
+    }
+
+    private void OnRemove(Entity<TurnstileComponent> turnstile, ref ComponentRemove args)
+    {
+        _activeTurnstiles.Remove(turnstile);
+    }
+
+    private bool GetIsCollidingWithAdmitted(Entity<TurnstileComponent, PreventCollideComponent?> ent)
+    {
+        if (!Resolve(ent.Owner, ref ent.Comp2))
+            return false;
+
+        var turnstileAABB = EntityLookupSystem.GetWorldAABB(ent.Owner);
+        var otherAABB = EntityLookupSystem.GetWorldAABB(ent.Comp2!.Uid);
+
+        if (turnstileAABB.Intersects(otherAABB))
+            return true;
+
+        return false;
+    }
+
+    /// <summary>
+    ///     Iterate over active turnstiles and progress them to the next state if they need to be updated.
+    /// </summary>
+    public override void Update(float frameTime)
+    {
+        var time = GameTiming.CurTime;
+        foreach (var ent in _activeTurnstiles.ToList())
+        {
+            SetCollidable(ent, true);
+            var turnstile = ent.Comp;
+            if (turnstile.Deleted || turnstile.NextStateChange == null)
+            {
+                _activeTurnstiles.Remove(ent);
+                continue;
+            }
+
+            if (Paused(ent))
+                continue;
+
+            if (turnstile.NextStateChange.Value < time)
+                NextState(ent, time);
+        }
+    }
```

Comment:
As I've mentioned elsewhere turnstiles should either be using an edge shape or a joint, but this PR just copies door code and tries to add the functionality on top which shouldn't be done.

Reply:
I thought you mentioned in #16419 that you would have preferred it to use StartCollideEvent & PreventCollide - I was intending to follow the same course of action. 
Apologies that this PR doesn't meet the standards required. Much of the code is inspired by the door code, but I tried to keep it minimal. 
I'm happy to make additional changes - which direction would you prefer the code to take?

Reply:
Made some changes, minifying the code and removing raised events that nothing was listening to. Please let me know if you still want the turnstile to use an edge shape or a dynamics joint, rather than the StartCollideEvent & PreventCollide combination.